pkg/ha/state/state.go

```diff
@@ -27,58 +45,116 @@ type StateView struct {
 	RecentLeaderWriteTime time.Time
 }
 
-func (s StateView) String() string {
-	return fmt.Sprintf("{\n\tLead:%s;\n\tUntil:%s;\n\tMaxSeen:%s;\n\tMaxInst:%s;\n\tMaxSeenLeader:%s\n}",
-		s.Leader, s.LeaseUntil, s.MaxTimeSeen, s.MaxTimeInstance, s.MaxTimeSeenLeader)
-}
+// Creates a new Lease and immediately synchronizes with the database, it either
+//	- sets the potentialLeader as the leader for the cluster with a lease
+//	  for the requested minT and maxT
+//	- or the existing leader and lease details are returned and set in the
+//	  new Lease.
+// An error is returned if an error occurred querying the database.
+func NewLease(c client.LeaseClient, cluster, potentialLeader string, minT, maxT, currentTime time.Time) (*Lease, error) {
+	dbState, err := c.UpdateLease(context.Background(), cluster, potentialLeader, minT, maxT)
+	if err != nil {
+		return nil, fmt.Errorf("could not create new lease: %#v", err)
+	}
+	var maxTimeInstance string
+	if maxT.After(dbState.LeaseUntil) {
+		maxTimeInstance = potentialLeader
+	} else {
+		maxTimeInstance = dbState.Leader
+	}
```

Comment:
Why is it wrong to always use potentialLeade for maxTimeInstance? I mean the maxTimeSeen will be whatever the potentialLeader sent. If there is a good reason let's add a comment.

Reply:
I forgot that if leader && maxT > leaseUntil the lease would be extended in the db call